### PR TITLE
Replace InMemoryPersistedState with GatewayMetaState in CoordinatorTests

### PR DESCRIPTION
**Pull request by andrershov**
This PR replaces `InMemoryPersistedState` with`GatewayMetaState` in
`CoordinatorTests`.
When performing such a replacement the main question was: do we want to
emulate exceptions as we do today in `MockPersistedState` before
delegating to `GatewayMetaState` or do we want these exceptions to
propagate from the lower level, i.e. file system exceptions?
On the one hand, lower level exception propagation is already tested in
`GatewayMetaStateTests`, so this won't improve the coverage. 
On the other hand, the benefit of low-level exceptions is to see how all these
components work in conjunction.  Finally, we abandoned the idea of low-level
exceptions because we don't have a way to deal with `IOError` today in
`CoordinatorTests`, but hacking `GatewayMetaState` not to throw
`IOError` seems unnatural.
So `MockPersistedState` rarely throws an exception before delegating to
`GatewayMetaState`, which is not supposed to throw the exception.

This commit required two changes:
1. Move `GatewayMetaStateUT` to upper-level from
`GatewayMetaStatePersistedStateTests`, because otherwise it's not easy
to construct `GatewayMetaState` instance in `CoordinatorTests`.
2. Move addition of `STATE_NOT_RECOVERED_BLOCK` from `GatewayMetaState`
constructor to `GatewayMetaState.applyClusterUpdaters`, because
`CoordinatorTests` class assume that there is no such block and most of
them fail.

Other classes can be migrated from `InMemoryPersistedState` to
`GatewayMetaState` as well if requested by reviewers.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateUT.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * {@link GatewayMetaState} constructor accepts a lot of arguments.
+ * It's not always easy / convenient to construct these dependencies.
+ * This class constructor takes far fewer dependencies and constructs usable {@link GatewayMetaState} with 2 restrictions:
+ * no metadata upgrade will be performed and no cluster state updaters will be run. This is sufficient for most of the tests.
+ * Metadata upgrade is tested in {@link  GatewayMetaStateTests} and different {@link ClusterStateUpdaters} in
+ * {@link ClusterStateUpdatersTests}.
+ */
+public class GatewayMetaStateUT extends GatewayMetaState {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateUT.java]: Maybe `MockGatewayMetaState` would be a better name?

File: server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateUT.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    private final DiscoveryNode localNode;
+
+    public GatewayMetaStateUT(Settings settings, NodeEnvironment nodeEnvironment,
+                              NamedXContentRegistry xContentRegistry,
+                       DiscoveryNode localNode) throws IOException {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateUT.java]: nit: funny indentation.

**Review by DaveCTurner — COMMENTED** [review]
Looks good. I suggested a renaming and a nit. I'm going to run this for a few iterations on my CI machine to be sure before we merge it.

**Review by DaveCTurner — COMMENTED** [review]
Hmm, this change _substantially_ adds to the time it takes to run these tests. I guess this is to be expected since we're hitting the disk and fsyncing and so on. I ran Gradle with these options:

```
:server:test -Dtests.class=org.elasticsearch.cluster.coordination.CoordinatorTests -Dtests.jvm.argline=-Dhppc.bitmixer=DETERMINISTIC -Dtests.iters=20 -Dtests.failfast=true
```

On `master`, the second time through this took 514 sec. On this branch, it took 891 sec, 73% longer. Maybe we should only use a proper persistence layer rarely?

**Comment by ywelsch** [comment]
+1 on using it rarely

**Comment by andrershov** [comment]
@DaveCTurner I've addressed your comments and changed `MockPersistedState` to use `GatewayMetaState` rarely and use `InMemoryPersistedState` in other cases. Ready for the 2nd pass.

**Comment by andrershov** [comment]
run gradle build tests 1

**Comment by andrershov** [comment]
run gradle build tests 2

**Review by DaveCTurner — COMMENTED** [review]
Just a couple more nits.

**Comment by andrershov** [comment]
@DaveCTurner Can you please make the next pass? I've fixed nits and last merge with master introduced node restart in CoordinatorTests (https://github.com/elastic/elasticsearch/pull/36897/commits/40b303cdd915098235704064ff844f59f25e2abc), so now `MockPersistedState` supports restarts when `GatewayMetaState` is used.

**Review by DaveCTurner — APPROVED** [review]
LGTM, thanks @andrershov. Could you add a version label before merging?

**Comment by andrershov** [comment]
run gradle build tests

**Comment by andrershov** [comment]
test this please

**Comment by andrershov** [comment]
run gradle build tests 2

**Pull request merged**